pkg/codegen/schema/bind.go

```diff
@@ -1476,40 +1476,125 @@ func (t *types) bindFunctionDef(token string) (*Function, hcl.Diagnostics, error
 		diags = diags.Extend(inDiags)
 		if err != nil {
 			return nil, diags, fmt.Errorf("error binding inputs for function %v: %w", token, err)
 		}
+
+		if len(spec.MultiArgumentInputs) > 0 {
+			idx := make(map[string]int, len(spec.MultiArgumentInputs))
+			for i, k := range spec.MultiArgumentInputs {
+				idx[k] = i
+			}
+			// Check that MultiArgumentInputs matches up 1:1 with the input properties
+			for k, i := range idx {
+				if _, ok := spec.Inputs.Properties[k]; !ok {
+					diags = diags.Append(&hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  fmt.Sprintf("multiArgumentInputs[%d] refers to non-existent property %#v", i, k),
+					})
+				}
+			}
+			var detailGiven bool
+			for k := range spec.Inputs.Properties {
+				if _, ok := idx[k]; !ok {
+					diag := hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  fmt.Sprintf("Property %#v not specified by multiArgumentInputs", k),
+					}
+					if !detailGiven {
+						detailGiven = true
+						diag.Detail = "If multiArgumentInputs is given, all properties must be specified"
+					}
+					diags = diags.Append(&diag)
+				}
+			}
+
+			// Order output properties as specified by MultiArgumentInputs
+			sortProps := func(props []*Property) {
+				sort.Slice(props, func(i, j int) bool {
+					return idx[props[i].Name] < idx[props[j].Name]
+				})
+			}
+			sortProps(ins.Properties)
+			if ins.InputShape != nil {
+				sortProps(ins.InputShape.Properties)
+			}
+			if ins.PlainShape != nil {
+				sortProps(ins.PlainShape.Properties)
+			}
+		}
+
 		inputs = ins
 	}
 
 	var outputs *ObjectType
-	if spec.Outputs != nil {
+
+	language := make(map[string]interface{})
+	for name, raw := range spec.Language {
+		language[name] = json.RawMessage(raw)
+	}
+
+	var inlineObjectAsReturnType bool
+	var returnType Type
+	if spec.ReturnType != nil && spec.Outputs == nil {
+		// compute the return type from the spec
+		if spec.ReturnType.ObjectTypeSpec != nil {
+			// bind as an object type
+			outs, outDiags, err := t.bindAnonymousObjectType(path+"/outputs", token+"Result", *spec.Outputs)
+			diags = diags.Extend(outDiags)
+			if err != nil {
+				return nil, diags, fmt.Errorf("error binding outputs for function %v: %w", token, err)
+			}
+			returnType = outs
+			outputs = outs
+			inlineObjectAsReturnType = true
+		} else if spec.ReturnType.TypeSpec != nil {
+			out, outDiags, err := t.bindTypeSpec(path+"/outputs", *spec.ReturnType.TypeSpec, false)
+			diags = diags.Extend(outDiags)
+			if err != nil {
+				return nil, diags, fmt.Errorf("error binding outputs for function %v: %w", token, err)
+			}
+			returnType = out
+		} else {
+			// Setting `spec.ReturnType` to a value without setting either `TypeSpec` or `ObjectTypeSpec`
+			// indicates a logical bug in our marshaling code.
+			return nil, diags, fmt.Errorf("error binding outputs for function %v: invalid return type", token)
+		}
+	} else if spec.Outputs != nil {
```

Comment:
@iwahbe This check here is the main difference. This is the new code block
```go
if spec.ReturnType != nil && spec.Outputs == nil {
  // bind from return type
} else if spec.Outputs != nil { 
  // bind from spec.Outputs
}
```
whereas before it was only:
```go
if spec.ReturnType != nil  {
  // bind from return type
}
```
Which caused the issue that @danielrbradley had where `spec.Outputs` was specified by code, not through JSON/YAML parsing and didn't bind correctly. 

I've also added a test to show that this actually works: `Function.ReturnType` being bound when `FunctionSpec.Outputs` is specified.